python/lsst/daf/butler/core/_butlerUri.py

```diff
@@ -0,0 +1,1332 @@
+        else:
+            # Use local file and upload it
+            local_src, is_temporary = src.as_local()
+
+            # resource.meta.upload_file seems like the right thing
+            # but we have a low level client
+            with open(local_src, "rb") as fh:
+                self.client.put_object(Bucket=self.netloc,
+                                       Key=self.relativeToPathRoot, Body=fh)
+            if is_temporary:
+                os.remove(local_src)
+
+        # This was an explicit move requested from a remote resource
+        # try to remove that resource
+        if transfer == "move":
+            # Transactions do not work here
+            src.remove()
+
+
+class ButlerPackageResourceURI(ButlerURI):
+    """URI referring to a Python package resource.
+
+    These URIs look like: ``resource://lsst.daf.butler/configs/file.yaml``
+    where the network location is the Python package and the path is the
+    resource name.
+    """
+
+    def exists(self) -> bool:
+        """Check that the python resource exists."""
+        return pkg_resources.resource_exists(self.netloc, self.relativeToPathRoot)
+
+    def read(self, size: int = -1) -> bytes:
+        with pkg_resources.resource_stream(self.netloc, self.relativeToPathRoot) as fh:
+            return fh.read(size)
+
+
+class ButlerHttpURI(ButlerURI):
+    """General HTTP(S) resource."""
+
+    def exists(self) -> bool:
+        """Check that a remote HTTP resource exists."""
+        header = requests.head(self.geturl())
+        return True if header.status_code == 200 else False
+
+    def as_local(self) -> Tuple[str, bool]:
+        """Download object over HTTP and place in temporary directory.
+
+        Returns
+        -------
+        path : `str`
+            Path to local temporary file.
+        temporary : `bool`
+            Always returns `True`. This is always a temporary file.
+        """
+        r = requests.get(self.geturl(), stream=True)
+        if r.status_code != 200:
+            raise FileNotFoundError(f"Unable to download resource {self}; status code: {r.status_code}")
+        with tempfile.NamedTemporaryFile(suffix=self.getExtension(), delete=False) as tmpFile:
+            for chunk in r.iter_content():
+                tmpFile.write(chunk)
+        return tmpFile.name, True
+
+    def read(self, size: int = -1) -> bytes:
+        # Docstring inherits
+        stream = True if size > 0 else False
+        r = requests.get(self.geturl(), stream=stream)
+        if not stream:
+            return r.content
+        else:
+            return next(r.iter_content(chunk_size=size))
+
+
+class ButlerGenericURI(ButlerURI):
+    """Generic URI with a defined scheme"""
+
+    def exists(self) -> bool:
+        """Test for existence and always return False."""
+        return False
+
+    def as_local(self) -> Tuple[str, bool]:
+        raise RuntimeError(f"Do not know how to retrieve data for URI '{self}'")
```

Comment:
What is the purpose of this class as opposed to throwing an exception from `ButlerURI.__new__`? That should ideally be in the documentation.

Reply:
It existed when I had abstractmethod in the base class but I didn't want to break everything up front if someone used an unsupported URI scheme. In theory the "PurePath" portion of ButlerURI is fine with an unknown scheme, it should only complain if you want to do something with an actual remote resource.

Reply:
Right, but do you ever want to use it that way? There is a bit of safety benefit to just throwing an exception in `__new__` when given an unsupported URL. For example, you will catch and throw a clear exception for `htpt` URLs and other typos, whereas now I think you may end up with obscure `NotImplementedError` exceptions much later on.

Reply:
Ok. I'll try the extreme prejudice approach and see what happens.